mythril_core/src/ioapic.rs

```diff
@@ -0,0 +1,677 @@
+            0b100 => Ok(DeliveryMode::NMI),
+            0b101 => Ok(DeliveryMode::INIT),
+            0b111 => Ok(DeliveryMode::ExtINT),
+            _ => Err(Error::InvalidValue(format!(
+                "Invalid pin polarity: 0x{:x}",
+                value
+            ))),
+        }
+    }
+}
+
+/// The status of the delivery of this interrupt.
+#[repr(u8)]
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum DeliveryStatus {
+    /// No activity on this interrupt.
+    Idle = 0x00,
+    /// Interrupt has been injected, but is currently pending.
+    SendPending = 0x01,
+}
+
+impl TryFrom<u8> for DeliveryStatus {
+    type Error = Error;
+
+    fn try_from(value: u8) -> Result<DeliveryStatus> {
+        match value {
+            0x00 => Ok(DeliveryStatus::Idle),
+            0x01 => Ok(DeliveryStatus::SendPending),
+            _ => Err(Error::InvalidValue(format!(
+                "Invalid delivery status: 0x{:x}",
+                value
+            ))),
+        }
+    }
+}
+
+/// A entry in the I/O Redirection Table.
+#[derive(Debug, Clone)]
+pub struct IoRedTblEntry {
+    /// The interrupt vector.
+    vector: u8,
+    /// The action the APIC should take on signal.
+    delivery_mode: DeliveryMode,
+    /// The interpretation of the destination field.
+    destination_mode: DestinationMode,
+    /// The current status of the delivery of this interrupt.
+    delivery_status: DeliveryStatus,
+    /// Polarity of the input signal.
+    pin_polarity: PinPolarity,
+    /// This is used for level triggered interrupts, this bit is set
+    /// to `true` when local APIC(s) accept the level interrupt sent by the
+    /// IOAPIC. The Remote IRR bit is set to 0 when an EOI message with
+    /// a matching interrupt vector is received from a local APIC.
+    remote_irr: bool,
+    /// Type of singal on interrupt pin.
+    trigger_mode: TriggerMode,
+    /// The interrupt signal is masked.
+    interrupt_mask: bool,
+    /// The destination field may be a logical set of processors or an
+    /// APIC ID.
+    /// let
+    destination: u8,
+}
+
+impl IoRedTblEntry {
+    /// Create a IO Redirection Table Entry.
+    pub fn new(
+        vector: u8,
+        delivery_mode: DeliveryMode,
+        destination_mode: DestinationMode,
+        pin_polarity: PinPolarity,
+        trigger_mode: TriggerMode,
+        interrupt_mask: bool,
+        destination: u8,
+    ) -> Result<IoRedTblEntry> {
+        let entry = IoRedTblEntry {
+            vector,
+            delivery_mode,
+            destination_mode,
+            delivery_status: DeliveryStatus::Idle,
+            pin_polarity,
+            remote_irr: false,
+            trigger_mode,
+            interrupt_mask,
+            destination,
+        };
+
+        entry.is_valid()?;
+
+        Ok(entry)
+    }
+
+    /// Perform basic validity checks found in the table from section 3.2.4
+    /// in the I/O APIC specification.
+    fn is_valid(&self) -> Result<()> {
```

Comment:
I understand the value of having some context for the error, so we don't want to just return bool for this. But given this signature, I'd like it to be called something like `validate`.